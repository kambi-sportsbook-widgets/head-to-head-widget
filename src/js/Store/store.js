import { coreLibrary, widgetModule, translationModule, statisticsModule } from 'kambi-widget-core-library';

/**
 * Builds statistics information object for usage in template.
 * @param {object} statistics Statistics entity
 * @returns {object[]}
 */
const parseDataInfo = function (statistics) {
   var dataInfo = [];
   var lastEvents = [];
   var allScores = [];
   var homeTeam = statistics.homeParticipant.participantName;
   var awayTeam = statistics.awayParticipant.participantName;

   statistics.lastEvents.forEach(function (lastEvent) {
      // Creating the item with the relevant data for each event
      const item = {
         start: lastEvent.start,
         homeParticipant: lastEvent.homeParticipant.participantName,
         awayParticipant: lastEvent.awayParticipant.participantName,
         homeScore: lastEvent.scores[0].homeScore,
         awayScore: lastEvent.scores[0].awayScore,
         type: lastEvent.scores[0].type
      };

      // Matching the scores of previous events with the current home team
      if (homeTeam === item.homeParticipant) {
         item.scoreOnLeft = item.homeScore;
         item.scoreOnRight = item.awayScore;
      } else {
         item.scoreOnLeft = item.awayScore;
         item.scoreOnRight = item.homeScore;
      }

      for (let j = 0; j < lastEvent.scores.length; j++) {
         if (lastEvent.scores[j].hasOwnProperty('homeScore') && lastEvent.scores[j].hasOwnProperty('awayScore')) {
            allScores.push(lastEvent.scores[j].homeScore, lastEvent.scores[j].awayScore);
         }
      }

      if (lastEvent.scores.length && lastEvent.scores[0].hasOwnProperty('homeScore') && lastEvent.scores[0].hasOwnProperty('awayScore')) {
         lastEvents.push(item);
      }
   });

   dataInfo.push({
      eventId: statistics.eventId,
      homeTeam: homeTeam,
      awayTeam: awayTeam,
      lastEvents: lastEvents,
      maxScore: Math.max.apply(null, allScores)
   });

   return dataInfo;
};

const getEventStatistics = function (eventId) {
   if (!eventId) {
      console.warn('eventId set from pageParam');
      eventId = coreLibrary.pageInfo.pageParam;
   } else {
      console.warn('eventId set from args.eventId');
   }

   return statisticsModule.getHeadToHeadStatistics(eventId)
   .then(parseDataInfo);
};

export default { getEventStatistics };
