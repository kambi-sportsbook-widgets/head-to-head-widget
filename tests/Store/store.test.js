import store from '../../src/js/Store/store';
import { statisticsModule } from 'kambi-widget-core-library';

jest.mock('kambi-widget-core-library', () => ({
   coreLibrary: {
      pageInfo: {
         pageParam: 0xfaff
      }
   },
   statisticsModule: {
      getHeadToHeadStatistics: jest.fn()
   }
}));

const mockStatistics = {
   homeParticipant: {
      participantName: 'Poland'
   },
   awayParticipant: {
      participantName: 'Sweden'
   },
   lastEvents: []
};

const mockEvent = {
   start: 100,
   homeParticipant: { participantName: 'Poland' },
   awayParticipant: { participantName: 'Sweden' },
   scores: [
      { homeScore: 1, awayScore: 0, type: 'TYPE_A' }
   ]
};

const mockEventSwapped = {
   start: 101,
   homeParticipant: { participantName: 'Sweden' },
   awayParticipant: { participantName: 'Poland' },
   scores: [
      { homeScore: 0, awayScore: 1, type: 'TYPE_A' }
   ]
};

const mockEventWithMissingAwayScore = {
   start: 102,
   homeParticipant: { participantName: 'Poland' },
   awayParticipant: { participantName: 'Sweden' },
   scores: [
      { homeScore: 1, type: 'TYPE_A' }
   ]
};

const mockEventWithMissingHomeScore = {
   start: 102,
   homeParticipant: { participantName: 'Poland' },
   awayParticipant: { participantName: 'Sweden' },
   scores: [
      { awayScore: 0, type: 'TYPE_A' }
   ]
};

describe('store.getEventStatistics()', () => {

   afterEach(() => {
      statisticsModule.getHeadToHeadStatistics.mockClear();
   });

   it('reads eventId from pageParam', () => {
      statisticsModule.getHeadToHeadStatistics = jest.fn(eventId =>
         new Promise(resolve => resolve(Object.assign(mockStatistics, { eventId }))));

      return store.getEventStatistics()
         .then((result) => {
            expect(result).toMatchSnapshot();
            expect(statisticsModule.getHeadToHeadStatistics).toHaveBeenCalledTimes(1);
         });
   });

   it('reads eventId from arguments', () => {
      statisticsModule.getHeadToHeadStatistics = jest.fn(eventId =>
         new Promise(resolve => resolve(Object.assign(mockStatistics, { eventId }))));

      return store.getEventStatistics(1001)
         .then((result) => {
            expect(result).toMatchSnapshot();
            expect(statisticsModule.getHeadToHeadStatistics).toHaveBeenCalledTimes(1);
         });
   });

   it('returns proper statistics', () => {
      statisticsModule.getHeadToHeadStatistics = jest.fn(eventId =>
         new Promise(resolve => resolve(Object.assign(
            mockStatistics,
            { eventId },
            { lastEvents: [mockEvent] }
         ))));

      return store.getEventStatistics(1002)
         .then((result) => {
            expect(result).toMatchSnapshot();
            expect(statisticsModule.getHeadToHeadStatistics).toHaveBeenCalledTimes(1);
         });
   });

   it('swaps teams to match home/away order', () => {
      statisticsModule.getHeadToHeadStatistics = jest.fn(eventId =>
         new Promise(resolve => resolve(Object.assign(
            mockStatistics,
            { eventId },
            { lastEvents: [mockEventSwapped] }
         ))));

      return store.getEventStatistics(1003)
         .then((result) => {
            expect(result).toMatchSnapshot();
            expect(statisticsModule.getHeadToHeadStatistics).toHaveBeenCalledTimes(1);
         });
   });

   it('does not include scores with missing home score', () => {
      statisticsModule.getHeadToHeadStatistics = jest.fn(eventId =>
         new Promise(resolve => resolve(Object.assign(
            mockStatistics,
            { eventId },
            { lastEvents: [mockEventWithMissingHomeScore] }
         ))));

      return store.getEventStatistics(1004)
         .then((result) => {
            expect(result).toMatchSnapshot();
            expect(statisticsModule.getHeadToHeadStatistics).toHaveBeenCalledTimes(1);
         });
   });

   it('does not include scores with missing away score', () => {
      statisticsModule.getHeadToHeadStatistics = jest.fn(eventId =>
         new Promise(resolve => resolve(Object.assign(
            mockStatistics,
            { eventId },
            { lastEvents: [mockEventWithMissingAwayScore] }
         ))));

      return store.getEventStatistics(1005)
         .then((result) => {
            expect(result).toMatchSnapshot();
            expect(statisticsModule.getHeadToHeadStatistics).toHaveBeenCalledTimes(1);
         });
   });

});
