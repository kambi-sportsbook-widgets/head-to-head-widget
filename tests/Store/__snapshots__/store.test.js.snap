// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store.getEventStatistics() does not include scores with missing away score 1`] = `
Array [
  Object {
    "awayTeam": "Sweden",
    "eventId": 1005,
    "homeTeam": "Poland",
    "lastEvents": Array [],
    "maxScore": -Infinity,
  },
]
`;

exports[`store.getEventStatistics() does not include scores with missing home score 1`] = `
Array [
  Object {
    "awayTeam": "Sweden",
    "eventId": 1004,
    "homeTeam": "Poland",
    "lastEvents": Array [],
    "maxScore": -Infinity,
  },
]
`;

exports[`store.getEventStatistics() reads eventId from arguments 1`] = `
Array [
  Object {
    "awayTeam": "Sweden",
    "eventId": 1001,
    "homeTeam": "Poland",
    "lastEvents": Array [],
    "maxScore": -Infinity,
  },
]
`;

exports[`store.getEventStatistics() reads eventId from pageParam 1`] = `
Array [
  Object {
    "awayTeam": "Sweden",
    "eventId": 64255,
    "homeTeam": "Poland",
    "lastEvents": Array [],
    "maxScore": -Infinity,
  },
]
`;

exports[`store.getEventStatistics() returns proper statistics 1`] = `
Array [
  Object {
    "awayTeam": "Sweden",
    "eventId": 1002,
    "homeTeam": "Poland",
    "lastEvents": Array [
      Object {
        "awayParticipant": "Sweden",
        "awayScore": 0,
        "homeParticipant": "Poland",
        "homeScore": 1,
        "scoreOnLeft": 1,
        "scoreOnRight": 0,
        "start": 100,
        "type": "TYPE_A",
      },
    ],
    "maxScore": 1,
  },
]
`;

exports[`store.getEventStatistics() swaps teams to match home/away order 1`] = `
Array [
  Object {
    "awayTeam": "Sweden",
    "eventId": 1003,
    "homeTeam": "Poland",
    "lastEvents": Array [
      Object {
        "awayParticipant": "Poland",
        "awayScore": 1,
        "homeParticipant": "Sweden",
        "homeScore": 0,
        "scoreOnLeft": 1,
        "scoreOnRight": 0,
        "start": 101,
        "type": "TYPE_A",
      },
    ],
    "maxScore": 1,
  },
]
`;
